### PR TITLE
embed: Delay setting initial cluster

### LINKED ISSUE #7516: Multiple discovery flags error with valid config file
A valid config file triggers a config verification bug about multiple discovery/boot strap flags.

In my example, I'm clearly **only** specifying `discovery-srv`.

Bug is present in latest release (3.1.3) and also with all the older 3.0 releases I've tried (3.0.15, 3.0.17).

## Example

Given a config file:

`config.json`:
```json
{
    "advertise-client-urls": "http://etcd.test:2379",
    "discovery-srv": "test",
    "initial-advertise-peer-urls": "http://etcd.test:2380",
    "initial-cluster-state": "new",
    "initial-cluster-token": "token",
    "listen-client-urls": "http://etcd1.test:2379",
    "listen-peer-urls": "http://etcd1.test:2380",
    "name": "etcd1"
}
```

And trying to start etcd:

```
> etcd-v3.1.3-darwin-amd64/etcd --config-file config.json
2017-03-16 12:05:21.978435 I | etcdmain: Loading server configuration from "config.json"
2017-03-16 12:05:21.978858 W | embed: expected IP in URL for binding (http://etcd1.test:2380)
2017-03-16 12:05:21.978869 W | embed: expected IP in URL for binding (http://etcd1.test:2379)
2017-03-16 12:05:21.978875 E | etcdmain: error verifying flags, multiple discovery or bootstrap flags are set. Choose one of "initial-cluster", "discovery" or "discovery-srv". See 'etcd --help'.
```

The equivalent invocation using CLI flags works just fine however:

```
./etcd-v3.1.3-darwin-amd64/etcd \
    --name "etcd1" \
    --discovery-srv "test" \
    --advertise-client-urls "http://etcd.test:2379" \
    --initial-advertise-peer-urls "http://etcd.test:2380" \
    --initial-cluster-state "new" \
    --initial-cluster-token "token" \
    --listen-client-urls "http://etcd1.test:2379" \
    --listen-peer-urls "http://etcd1.test:2380"
```

## Hack fix

The only way to get etcd to accept this config is by adding an empty `initial-cluster` string, which contradicts the error message (I have `initial-cluster` **and** `discovery-srv` specified).

`hacked-config.json`:
```json
{
    "advertise-client-urls": "http://etcd.test:2379",
    "discovery-srv": "test",
    "initial-advertise-peer-urls": "http://etcd.test:2380",
    "initial-cluster": "",
    "initial-cluster-state": "new",
    "initial-cluster-token": "token",
    "listen-client-urls": "http://etcd1.test:2379",
    "listen-peer-urls": "http://etcd1.test:2380",
    "name": "etcd1"
}
```

and now etcd will start:

```
> etcd-v3.1.3-darwin-amd64/etcd --config-file config.json
2017-03-16 12:07:55.997893 I | etcdmain: Loading server configuration from "config.json"
2017-03-16 12:07:55.998342 W | embed: expected IP in URL for binding (http://etcd1.test:2380)
2017-03-16 12:07:55.998353 W | embed: expected IP in URL for binding (http://etcd1.test:2379)
2017-03-16 12:07:55.998477 I | etcdmain: etcd Version: 3.1.3
2017-03-16 12:07:55.998483 I | etcdmain: Git SHA: 21fdcc6
2017-03-16 12:07:55.998486 I | etcdmain: Go Version: go1.7.5
2017-03-16 12:07:55.998488 I | etcdmain: Go OS/Arch: darwin/amd64
2017-03-16 12:07:55.998493 I | etcdmain: setting maximum number of CPUs to 8, total number of available CPUs is 8
2017-03-16 12:07:55.998498 N | etcdmain: failed to detect default host (default host not supported on darwin_amd64)
2017-03-16 12:07:55.998504 W | etcdmain: no data-dir provided, using default data-dir ./etcd1.etcd
```

### PR DESCRIPTION
**Pull request by jsok**
PR #7517 attempted to address this however it only worked if the name was left as "default".

NewConfig() sets an initial cluster (potentially using a default name) but we should clear it in the event another discovery option has been specified.

(Completely) Fixes #7516

### CODE REVIEW

**Comment by heyitsanthony** [comment]
s/embed: Delay setting initial cluster/embed: Delay setting initial cluster for YAML/ for the first line of the commit message and this should be good to merge. Thanks!

**Comment by xiang90** [comment]
@jsok It would be great if you can contribute a test :).

**Comment by jsok** [comment]
@xiang90 sure thing, will append to PR soon.
@heyitsanthony thanks, will re-word the commit message.

**Comment by jsok** [comment]
@xiang90 @heyitsanthony commit name amended and test case added!

File: etcdmain/config_test.go [hunk 1]
```diff
@@ -145,6 +145,62 @@ func TestConfigFileClusteringFields(t *testing.T) {
 	validateClusteringFlags(t, cfg)
 }
 
+func TestConfigFileClusteringFlags(t *testing.T) {
+	tests := []struct {
+		Name           string `json:"name"`
```
Comment (xiang90) [inline etcdmain/config_test.go]: use private field? name instead of Name?

Comment (xiang90) [inline etcdmain/config_test.go]: oh... we call marshal on it. nvm.

File: etcdmain/config_test.go [hunk 2]
```diff
@@ -145,6 +145,62 @@ func TestConfigFileClusteringFields(t *testing.T) {
...
+		InitialCluster string `json:"initial-cluster"`
+		DNSCluster     string `json:"discovery-srv"`
+		Durl           string `json:"discovery"`
+	}{
+		{
+		/* Use default name and generate a default inital-cluster */
```
Comment (xiang90) [inline etcdmain/config_test.go]: use go style comment? // ... instead of /* ... */

File: etcdmain/config_test.go [hunk 3]
```diff
@@ -145,6 +145,62 @@ func TestConfigFileClusteringFields(t *testing.T) {
...
+	for i, tt := range tests {
+		b, err := yaml.Marshal(&tt)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		tmpfile := mustCreateCfgFile(t, b)
+		defer os.Remove(tmpfile.Name())
+
+		args := []string{
```
Comment (fanminshi) [inline etcdmain/config_test.go]: one liner `args := []string{fmt.Sprintf("--config-file=%s", tmpfile.Name())}`?

File: embed/config.go [hunk 4]
```diff
@@ -246,7 +248,7 @@ func (cfg *configYAML) configFromFile(path string) error {
 		cfg.ACUrls = []url.URL(u)
 	}
 
-	if (cfg.Durl != "" || cfg.DNSCluster != "") && cfg.InitialCluster == cfg.InitialClusterFromName(cfg.Name) {
+	if (cfg.Durl != "" || cfg.DNSCluster != "") && cfg.InitialCluster == defaultInitialCluster {
```
Comment (fanminshi) [inline embed/config.go]: a comment here explaining why we need this check will be helpful.

File: etcdmain/config_test.go (hunk 3 above)
Comment (jsok) [inline etcdmain/config_test.go]: Was duplicating existing test code, have condensed and updated all other occurrences.

File: etcdmain/config_test.go (hunk 2 above)
Comment (jsok) [inline etcdmain/config_test.go]: Done

File: embed/config.go (hunk 4 above)
Comment (jsok) [inline embed/config.go]: Done... although it's tricky to explain :-\

Not a fan of all the default values set by `NewConfig`, as it makes for more complicated undo/override code like this.

**Comment by fanminshi** [comment]
lgtm

**Comment by xiang90** [comment]
LGTM defer to @heyitsanthony

File: etcdmain/config_test.go [hunk 5]
```diff
@@ -145,6 +141,60 @@ func TestConfigFileClusteringFields(t *testing.T) {
...
+
+		tmpfile := mustCreateCfgFile(t, b)
+		defer os.Remove(tmpfile.Name())
+
+		args := []string{fmt.Sprintf("--config-file=%s", tmpfile.Name())}
+
+		cfg := newConfig()
+		if err := cfg.parse(args); err != nil {
+			t.Errorf("%d: err = %v", i, err)
+		}
```
Comment (heyitsanthony) [inline etcdmain/config_test.go]: there's nothing here to test that the parsed data matches some expected value?

Comment (jsok) [inline etcdmain/config_test.go]: It's not a parsing test, it's attempting to ensure that if independent discovery/bootstrap flags are correctly specified they *don't* falsely trigger an `embed.ErrConflictBootstrapFlags` error.

I'm assuming there are existing tests to validate parsing works?

Comment (heyitsanthony) [inline etcdmain/config_test.go]: Since it's dealing with overriding data, I thought there'd be some value checks, but I guess that's OK

**Comment by heyitsanthony** [comment]
lgtm

File: etcdmain/config_test.go (hunk 5 above)
Comment (jsok) [inline etcdmain/config_test.go]: :+1:

**Pull request merged**